app/workers/moderator/sink_articles_worker.rb

```diff
@@ -8,10 +8,7 @@ def perform(user_id)
       user = User.find_by(id: user_id)
       return unless user
 
-      articles = Article.where(user: user)
-      reactions = Reaction.where(reactable: articles)
-      new_score = reactions.sum(:points) + Reaction.where(reactable: user).sum(:points)
-      articles.update_all(score: new_score)
+      Article.where(user: user).each(&:update_score)
```

Comment:
the other pertinent question is whether 1000 articles is a normal amount or an outlier (I can check, looks like we might have 25 or so users with more than 1000 articles, and I didn't filter on draft/published state)

```
select count(id) from (
select users.id as id, count(articles.id) as article_count from users
inner join articles on articles.user_id = users.id
group by users.id
) as t where article_count > 1000;
``` 
gives 28 users with this many, and a select few with 10-20K:

```
select count(articles.id) as article_count 
from users
inner join articles on articles.user_id = users.id
group by users.id
order by article_count desc
limit 20;

article_count
--
20,306
18,194
16,500
15,464
6,333
5,990
5,102
4,661
4,516
3,368
2,988
2,496
2,376
2,265
1,897
1,728
1,687
1,674
1,669
1,533
```

I'll benchmark a few scorings locally to get a feeling for the time used - I'm definitely not opposed to breaking this into separate batches (there's already a background worker to rescore one article which both updates score and indexes to elasticsearch - it might be the case that we want an article bulk score calc worker that takes a batch of article ids and indexes them

The `Article#index_to_elasticsearch_inline` present in the `Articles::ScoreCalcWorker#perform` call is also missing from the moderator action here, I'm not sure I understand yet what the impact of changing _only_ the database copy of the article's score would be (does the feed rely on the es document or is that only for internal search results?)

Reply:
@djuber 

> there's already a background worker to rescore one article which both updates score and indexes to elasticsearch

Nice! Sounds like we could enqueue that one in the loop?

> does the feed rely on the es document or is that only for internal search results?

I want to say it's only being used for search, but for consistency's sake it's probably a good idea to refresh the Elasticsearch index unless it's causing performance issues.

Elasticsearch is going away reasonably soon, though — our initial efforts at moving search over to Postgres FTS have been quite impactful.

Reply:
Locally (which has performance characteristics different from prod for obvious reasons) I get about 10ms per article per call to update_score

```
[5] pry(main)> Benchmark.measure do
[5] pry(main)*   1000.times do  
[5] pry(main)*     Article.first.update_score    
[5] pry(main)*   end  
[5] pry(main)* end  
=> #<Benchmark::Tms:0x000055b1aef4c288
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=10.635379651968833,
 @stime=0.2575569999999999,
 @total=8.566616999999999,
 @utime=8.309059999999999>
```

Article.count is 27 in my test instance, taking each one by one gives about  the same time (on the order of 10ms for the scoring, including the blackbox calls)

```
[6] pry(main)> def update_many(times)
[6] pry(main)*   Benchmark.measure do
[6] pry(main)*     times.times do
[6] pry(main)*       Article.all.each(&:update_score)
[6] pry(main)*     end  
[6] pry(main)*   end  
[6] pry(main)* end  
=> :update_many
[7] pry(main)> 
[8] pry(main)> update_many(10)
=> #<Benchmark::Tms:0x000055b1b5a64c50
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=2.4595410070032813,
 @stime=0.05515100000000006,
 @total=1.9854370000000006,
 @utime=1.9302860000000006>
[9] pry(main)> update_many(100)
=> #<Benchmark::Tms:0x000055b1b5de0ca8
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=24.696464038046543,
 @stime=0.510732,
 @total=19.968933999999997,
 @utime=19.458201999999996>
```

![update_score](https://user-images.githubusercontent.com/1237369/111340573-c40dbf00-8646-11eb-8a14-1988d8e715ed.png)

About 45% of the time is spent in the spaminess calculation, 20% of the total time is checking the trusted role for the user inside the spaminess calculation, a small amount in the hotness calculation, 30% sum, and 17% update_columns.

I realized I have the null store cache (development env default setting) so this might be unrealistic (if redis is faster than the role lookup in the db is), however the memoization only works for true and not false the way it's written (common idiom for memoization is `@ivar || get_value` but that's poorly adapted to booleans).

Reply:
I put the memoization fix under https://github.com/forem/forem/issues/13006 and can get that done this week.

Reply:
Flame graphs FTW :D

I was going to write a RFC 0 to propose adding [rack-mini-profiler](https://github.com/MiniProfiler/rack-mini-profiler) in dev and in production 👀

Reply:
@djuber This looks great! I'm happy to merge as-is and get real production metrics on it, but I really appreciate your effort on making sure it shouldn't take too long to run for someone with a massive number of posts! 🎉

@rhymes I'd be really interested in that 🙂

Reply:
I used rbspy to generate the flamegraph and timings here (https://rbspy.github.io/) - since it's running separately there's no runtime/bundle dependency, it's just sampling the processes memory space directly. 

I'll go ahead and merge this and prepare the bulk update if it's applicable and necessary (not sure what a normal cutoff is for a sidekiq process, if we're looking at 1s we'll exceed it, if we're okay with 10s job runtimes this probably isn't a huge problem for most users).

Reply:
if we also "fix" the issue where this is bypassing the es doc reindexing (which the scorer job does after updating the db) it's possible the runtime will increase substantially, but I'm not expecting "minutes" with a db only change.

Reply:
@djuber thank you for the tip! Didn't know about rbspy, definitely going to install it :)